### PR TITLE
cmake: sync with `configure.py` (14/n)

### PR DESCRIPTION
**Pull request by tchaikov**
this is the 14rd changeset of a series which tries to give an overhaul to the CMake building system. this series has two goals:
  - to enable developer to use CMake for building scylla. so they can use tools (CLion for instance) with CMake integration for better developer experience
  - to enable us to tweak the dependencies in a simpler way. a well-defined cross module / subsystem dependency is a prerequisite for building this project with the C++20 modules.

this changeset includes following changes:

- build: cmake: promote add_scylla_test() to test/
- build: cmake: add all tests

### CODE REVIEW

**Comment by tchaikov** [comment]
@denesb hi Botond, could you please help review this change?

File: test/unit/CMakeLists.txt [hunk 1]
```diff
@@ -2,21 +2,32 @@ add_scylla_test(bptree_compaction_test
...
+  ARGS -c1 -m1G --count=4132 --iter=9
+  ARGS1 -c1 -m1G --count=27 --iter=312)
 add_scylla_test(cross_shard_barrier_test
   KIND UNIT)
 add_scylla_test(lsa_sync_eviction_test
-  KIND UNIT)
+  KIND UNIT
+  ARGS -c1 -m100M --count 10 --standard-object-size 3000000
+  ARGS1 -c1 -m100M --count 24000 --standard-object-size 2048
+  ARGS2 -c1 -m1G --count 4000000 --standard-object-size 128)
```
Comment (denesb) [inline test/unit/CMakeLists.txt:20]: I don't understand how this will work. Will the test be run with all 3 different args?

**Review by denesb — COMMENTED** [review]
Not sure if it is worth trying to set up CTEST to run our tests. We have the monster that is `test.py` already as a test runner and I don't see how we could even start to replicate that in CTEST/CMAKE. So I think CMake should just generate a command to invoke `test.py`, like `configure.py` does, e.g.: `{mode}-test`.

File: test/unit/CMakeLists.txt (hunk 1 above)
Comment (tchaikov) [reply test/unit/CMakeLists.txt:20]: yes. it will create three tests.

**Comment by tchaikov** [comment]
> Not sure if it is worth trying to set up CTEST to run our tests. We have the monster that is `test.py` already as a test runner and I don't see how we could even start to replicate that in CTEST/CMAKE. So I think CMake should just generate a command to invoke `test.py`, like `configure.py` does, e.g.: `{mode}-test`.

i am imagining a development iteration, 

1. modify the code
2. run a certain test to verify the modified code

with CTest integration, i think i can make the built executable as a dependency of the test, and implement the 2nd step with a single command.  if we want to implement this with the combo of `test.py` + generated `build.ninja`, we would need to use two commands: for building the executable and for running the test using the test driver.

another advantage of this layout is that it is test centric not tool centric. the test parameters are located right near the test, not being gathered together.

**Comment by denesb** [comment]
> i am imagining a development iteration,
> 
>     1. modify the code
> 
>     2. run a certain test to verify the modified code
> 
> 
> with CTest integration, i think i can make the built executable as a dependency of the test, and implement the 2nd step with a single command. if we want to implement this with the combo of `test.py` + generated `build.ninja`, we would need to use two commands: for building the executable and for running the test using the test driver.
> 
> another advantage of this layout is that it is test centric not tool centric. so a certain test's parameters are located right near the test, not being gathered together with other tests.

In the typical development iteration I always prefer to run the test manually because this allows me to pass custom parameters. If I want to run the entire test file I use `test.py`. I don't see where CTEST would fit in here as I don't think CTEST allows passing custom cmd-line args to the test.

We already have a way to store desired parameters the test runner should use for running the test. How this is implemented may not be ideal but duplicating it will just lead to differences creeping in. I think we should keep a single source of truth w.r.t. the test parameters.
Giving options to the developers is fine, but when it comes to test runners, it is best if CI/promotion and developers use the same one. Also I just don't see how CTEST could possibly support all the different weird tests we have. I see the value of CTEST as a test runner, not for running individual tests while developing. And as I don't think it will ever be able to be our test runner, I don't think we should bother trying to bring

**Comment by tchaikov** [comment]
> > i am imagining a development iteration,
> > ```
> > 1. modify the code
> > 
> > 2. run a certain test to verify the modified code
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > with CTest integration, i think i can make the built executable as a dependency of the test, and implement the 2nd step with a single command. if we want to implement this with the combo of `test.py` + generated `build.ninja`, we would need to use two commands: for building the executable and for running the test using the test driver.
> > another advantage of this layout is that it is test centric not tool centric. so a certain test's parameters are located right near the test, not being gathered together with other tests.
> 
> In the typical development iteration I always prefer to run the test manually because this allows me to pass custom parameters. If I want to run the entire test file I use `test.py`. I don't see where CTEST would fit in here as I don't think CTEST allows passing custom cmd-line args to the test.

`okay` i think it's possible to support customized argument specified at runtime. but two copies of desired argument are indeed not desirable.

> 
> We already have a way to store desired parameters the test runner should use for running the test. How this is implemented may not be ideal but duplicating it will just lead to differences creeping in. I think we should keep a single source of truth w.r.t. the test parameters. Giving options to the developers is fine, but when it comes to test runners, it is best if CI/promotion and developers use the same one. Also I just don't see how CTEST could possibly support all the different weird tests we have. 

it could. but yeah, it might end up with an even more weird thing.

> I see the value of CTEST as a test runner, not for running individual tests while developing. And as I don't think it will ever be able to be our test runner, I don't think we should bother trying to make it an incomplete imitation of `test.py`.

if it is not able to drive individual test, its value is more much less than it should have as a tool. so i will drop it from this changeset. this is but a side product created when i was reading `test.py` though.

**Comment by tchaikov** [comment]
changelog:

- dropped ctest related changes

**Review by denesb — APPROVED** [review]

**Comment by tchaikov** [comment]
[ttl_test.TestDistributedTTL.test_ttl_is_respected_on_repair[parallel_aggregates_enabled]](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/105/testReport/junit/ttl_test/TestDistributedTTL/Sanity_Tests___test_ttl_is_respected_on_repair_parallel_aggregates_enabled_/) =(

**Pull request merged**